# HealthConnect Appointment Analysis

### AnalystLab Africa — Experience Lab Internship Programme
### Week 4: HealthConnect Project Kickoff & Problem Understanding

**Project:** HealthConnect Appointment Data Analysis  
**Analyst:** Comfort Ogunyemi  
**Tools:** Python, Pandas, Power BI, Jupyter Notebook 

## 1. Project Overview

HealthConnect is a healthcare appointment platform that connects patients with healthcare providers.

This project focuses on analysing appointment data to understand patient attendance patterns, identify factors associated with missed appointments, and generate actionable insights that can help HealthConnect improve appointment attendance and operational efficiency.

## 2. Business Problem

Missed medical appointments (no-shows) can negatively affect healthcare providers by creating unused appointment slots, reducing operational efficiency, and limiting access to care for other patients.

The objective of this analysis is to use historical appointment data to identify patterns and factors associated with appointment no-shows and provide data-driven recommendations for reducing missed appointments.

## 3. Project Objectives

The analysis will aim to:

1. Understand the structure and characteristics of the HealthConnect appointment dataset.
2. Clean and prepare the data for analysis.
3. Analyse appointment attendance and no-show patterns.
4. Identify demographic, scheduling, and other factors associated with missed appointments.
5. Develop relevant KPIs to measure appointment performance.
6. Generate actionable insights and recommendations for reducing no-shows.

## 4. Data Analytics Track and Role

As a Data Analytics intern, my role in the HealthConnect project is to examine appointment data to understand patterns associated with patient attendance and missed appointments.

The analysis will focus on identifying relevant variables, assessing data quality, defining business questions, and proposing meaningful Key Performance Indicators (KPIs) that can support decision-making.

The findings from this initial analysis will provide a foundation for deeper analysis and solution development in subsequent weeks.

## 5. Project Resources

The following resources are relevant to the Data Analytics track:

1. **HealthConnect Appointment Dataset**  
   Contains fictional and anonymized appointment records, including patient demographics, appointment details, booking information, previous appointment history, reminder information, distance to the clinic, waiting time, and appointment outcomes.

2. **HealthConnect Data Dictionary**  
   Provides definitions and explanations of the variables contained in the appointment dataset.

These resources will be used to understand the structure and meaning of the available data and determine which variables are relevant to analysing appointment attendance and no-show patterns.

In [12]:
import pandas as pd

df = pd.read_excel(r"C:\Users\DELL\Documents\HealthConnect_Project\HealthConnect_Appointment_Data.xlsx")

df.head()

,appointment_id,patient_id,gender,age,age_group,appointment_type,booking_date,appointment_date,appointment_day,appointment_time,booking_lead_days,previous_appointments,previous_no_shows,reminder_sent,reminder_channel,distance_to_clinic_km,waiting_time_minutes,appointment_outcome
0,HC-00001,P-1613,Female,39,35-44,Follow-up,2025-02-06,2025-02-18,Tuesday,Afternoon,12,2,0,Yes,WhatsApp,19.3,29.0,No-Show
1,HC-00002,P-0813,Male,31,25-34,Specialist Consultation,2026-02-25,2026-02-27,Friday,Morning,2,6,0,Yes,SMS,14.3,42.0,Attended
2,HC-00003,P-1366,Female,50,45-54,General Consultation,2025-11-16,2025-12-24,Wednesday,Morning,38,5,1,Yes,SMS,11.4,11.0,No-Show
3,HC-00004,P-1031,Male,59,55-64,Follow-up,2025-07-18,2025-08-28,Thursday,Evening,41,3,1,Yes,SMS,7.4,35.0,Attended
4,HC-00005,P-1458,Female,34,25-34,Follow-up,2025-07-09,2025-08-25,Monday,Afternoon,47,3,1,Yes,Email,5.6,27.0,No-Show


In [13]:
df.shape

(5000, 18)

In [14]:
df.columns.tolist()

['appointment_id',
 'patient_id',
 'gender',
 'age',
 'age_group',
 'appointment_type',
 'booking_date',
 'appointment_date',
 'appointment_day',
 'appointment_time',
 'booking_lead_days',
 'previous_appointments',
 'previous_no_shows',
 'reminder_sent',
 'reminder_channel',
 'distance_to_clinic_km',
 'waiting_time_minutes',
 'appointment_outcome']

In [15]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 18 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   appointment_id         5000 non-null   str           
 1   patient_id             5000 non-null   str           
 2   gender                 5000 non-null   str           
 3   age                    5000 non-null   int64         
 4   age_group              5000 non-null   str           
 5   appointment_type       5000 non-null   str           
 6   booking_date           5000 non-null   datetime64[us]
 7   appointment_date       5000 non-null   datetime64[us]
 8   appointment_day        5000 non-null   str           
 9   appointment_time       5000 non-null   str           
 10  booking_lead_days      5000 non-null   int64         
 11  previous_appointments  5000 non-null   int64         
 12  previous_no_shows      5000 non-null   int64         
 13  reminder_sent 

In [16]:
df.isnull().sum()

appointment_id              0
patient_id                  0
gender                      0
age                         0
age_group                   0
appointment_type            0
booking_date                0
appointment_date            0
appointment_day             0
appointment_time            0
booking_lead_days           0
previous_appointments       0
previous_no_shows           0
reminder_sent               0
reminder_channel         1366
distance_to_clinic_km      90
waiting_time_minutes       60
appointment_outcome         0
dtype: int64

In [17]:
df.duplicated().sum()

np.int64(0)

## 6. Dataset Overview

The HealthConnect appointment dataset contains 5,000 appointment records and 18 variables.

The dataset contains information covering several aspects of the patient appointment journey, including:

- Patient demographics
- Appointment characteristics
- Booking information
- Previous appointment history
- Previous no-show history
- Reminder information
- Distance to the clinic
- Waiting time
- Appointment outcomes

The dataset contains a mixture of categorical, numerical, and date/time variables. The two date variables (`booking_date` and `appointment_date`) are stored as datetime values, while numerical variables include age, booking lead days, previous appointments, previous no-shows, distance to the clinic, and waiting time.

The `appointment_outcome` variable is particularly important because it can be used to distinguish between different appointment outcomes and investigate patterns associated with missed appointments.

In [18]:
df["appointment_outcome"].value_counts()

appointment_outcome
No-Show      2423
Attended     2314
Cancelled     263
Name: count, dtype: int64

In [19]:
df["reminder_sent"].value_counts()

reminder_sent
Yes    3634
No     1366
Name: count, dtype: int64

In [20]:
df["reminder_channel"].value_counts(dropna=False)

reminder_channel
SMS         2000
NaN         1366
WhatsApp    1101
Email        533
Name: count, dtype: int64

In [21]:
df["appointment_type"].value_counts()

appointment_type
General Consultation       2086
Follow-up                  1421
Specialist Consultation     900
Diagnostic Test             593
Name: count, dtype: int64

In [22]:
df["gender"].value_counts()

gender
Female               2488
Male                 2404
Prefer not to say     108
Name: count, dtype: int64

In [23]:
df["age_group"].value_counts()

age_group
65+      1241
35-44     819
55-64     800
45-54     793
25-34     783
18-24     564
Name: count, dtype: int64

In [24]:
df[
    [
        "age",
        "booking_lead_days",
        "previous_appointments",
        "previous_no_shows",
        "distance_to_clinic_km",
        "waiting_time_minutes"
    ]
].describe()

,age,booking_lead_days,previous_appointments,previous_no_shows,distance_to_clinic_km,waiting_time_minutes
count,5000.000000,5000.00000,5000.000000,5000.000000,4910.000000,4940.000000
mean,48.794800,29.63860,3.013800,0.544200,10.109572,24.189676
std,18.138547,17.39936,1.741211,0.746832,6.590030,10.863184
min,18.000000,0.00000,0.000000,0.000000,0.500000,2.000000
25%,33.000000,15.00000,2.000000,0.000000,5.300000,17.000000
50%,49.000000,30.00000,3.000000,0.000000,8.700000,24.000000
75%,64.000000,45.00000,4.000000,1.000000,13.500000,32.000000
max,80.000000,60.00000,11.000000,5.000000,45.000000,68.000000


In [25]:
print("Unique appointment IDs:", df["appointment_id"].nunique())
print("Unique patient IDs:", df["patient_id"].nunique())

Unique appointment IDs: 5000
Unique patient IDs: 1696


## 7. Initial Data Quality Assessment

An initial assessment was conducted to evaluate the completeness, consistency, uniqueness, and general structure of the HealthConnect appointment dataset.

### Dataset completeness

The dataset contains 5,000 records across 18 variables. Most variables are complete. Missing values were identified in three variables:

- `reminder_channel`: 1,366 missing values
- `distance_to_clinic_km`: 90 missing values
- `waiting_time_minutes`: 60 missing values

The missing values in `reminder_channel` require further investigation because they may be related to whether a reminder was sent. Therefore, they should not automatically be treated as erroneous or removed.

### Duplicate and identifier assessment

There are no completely duplicated records in the dataset.

All 5,000 `appointment_id` values are unique, indicating that each record represents a distinct appointment.

There are 1,696 unique `patient_id` values among the 5,000 appointments. This indicates that patients may have multiple appointments within the dataset, which is relevant when considering previous appointment history and patient attendance patterns.

### Data types

The dataset contains appropriate data types for the initial analysis:

- Numerical variables include age, booking lead days, previous appointments, previous no-shows, distance to clinic, and waiting time.
- Categorical variables include gender, age group, appointment type, appointment day, reminder information, and appointment outcome.
- `booking_date` and `appointment_date` are stored as datetime variables.

### Appointment outcomes

The appointment outcome variable contains three categories:

- No-Show: 2,423
- Attended: 2,314
- Cancelled: 263

The presence of both cancelled and no-show appointments is important because these outcomes represent different appointment behaviours and should be considered separately during subsequent analysis.

### Initial assessment conclusion

The dataset appears suitable for further exploratory analysis. However, the missing values identified above and the treatment of cancelled appointments will need to be considered carefully during subsequent stages of the project.

## 8. Variables Relevant to Appointment Attendance and No-Shows

Based on the initial review of the dataset, the following variables are considered potentially relevant to understanding appointment attendance and no-show patterns:

| Variable | Potential Relevance |
|---|---|
| `appointment_outcome` | Primary outcome variable for distinguishing attended, no-show, and cancelled appointments. |
| `appointment_type` | May help identify whether attendance patterns differ by type of appointment. |
| `age` / `age_group` | May help identify differences in attendance behaviour across age groups. |
| `gender` | May be useful for examining potential differences in attendance patterns. |
| `booking_lead_days` | May help determine whether the time between booking and appointment is associated with attendance. |
| `previous_appointments` | May provide an indication of a patient's previous engagement with the clinic. |
| `previous_no_shows` | May help determine whether previous no-show behaviour is associated with future no-shows. |
| `reminder_sent` | Can be used to investigate the relationship between appointment reminders and attendance. |
| `reminder_channel` | Can help assess whether attendance patterns differ by reminder channel. |
| `distance_to_clinic_km` | May help investigate whether distance from the clinic is associated with missed appointments. |
| `waiting_time_minutes` | May be useful for understanding whether waiting time is associated with appointment outcomes. |
| `appointment_day` | May help identify differences in attendance patterns across days of the week. |
| `appointment_time` | May help investigate whether appointment time is associated with attendance. |

## 9. Business Questions

The following business questions will guide the subsequent analysis of the HealthConnect appointment data:

1. **What proportion of scheduled appointments result in no-shows, and how does this vary across different appointment outcomes?**

2. **Is there a relationship between appointment reminders and patient attendance?**

3. **Does the type of reminder channel appear to be associated with different attendance outcomes?**

4. **Which patient demographic or appointment characteristics are associated with higher rates of no-shows?**

5. **Is the length of time between booking and the appointment associated with the likelihood of a patient missing the appointment?**

6. **Does previous no-show behaviour appear to be associated with future appointment no-shows?**

7. **Are appointment day and appointment time associated with differences in attendance patterns?**

8. **Is distance to the clinic associated with appointment no-shows?**

## 10. Proposed Key Performance Indicators (KPIs)

The following KPIs are proposed for subsequent stages of the HealthConnect project:

| KPI | Definition | Linked Business Question | Relevance |
|---|---|---|---|
| **No-Show Rate** | Percentage of scheduled appointments that result in a no-show. | What proportion of scheduled appointments result in no-shows? | Provides a high-level measure of the clinic's appointment attendance problem and can be used to monitor improvement over time. |
| **Attendance Rate** | Percentage of scheduled appointments that are attended. | What proportion of scheduled appointments result in attendance? | Measures successful appointment attendance and provides a complementary view of the no-show rate. |
| **Reminder Effectiveness Rate** | Attendance rate among appointments where a reminder was sent, compared with appointments where no reminder was sent. | Is there a relationship between appointment reminders and patient attendance? | Helps assess whether the reminder process may contribute to improved appointment attendance. |
| **No-Show Rate by Appointment Type** | Percentage of appointments resulting in a no-show for each appointment type. | Are some appointment types associated with higher no-show rates? | Helps identify appointment categories that may require targeted engagement or intervention. |
| **Repeat No-Show Rate** | Proportion of appointments resulting in a no-show among patients with previous no-show history. | Does previous no-show behaviour appear to be associated with future no-shows? | Helps identify whether previous attendance behaviour may be useful for targeted patient engagement strategies. |

## 11. Initial Analysis Approach

The analysis will be conducted in stages to ensure that the appointment data is properly understood and prepared before drawing conclusions.

### Stage 1: Data Understanding and Quality Assessment
- Review the dataset structure and data dictionary.
- Examine variable types and categories.
- Assess missing values and duplicate records.
- Check identifier uniqueness.
- Investigate potential data quality issues.

### Stage 2: Data Preparation
- Determine appropriate treatment for missing values.
- Review categorical values for consistency.
- Confirm that date and time variables are suitable for analysis.
- Determine how cancelled appointments should be treated when analysing attendance and no-shows.
- Create any required derived variables where appropriate.

### Stage 3: Exploratory Analysis
- Examine the distribution of appointment outcomes.
- Analyse no-show patterns across relevant demographic and appointment characteristics.
- Investigate the relationship between reminders and appointment outcomes.
- Examine the potential relationship between booking lead time, previous no-shows, distance, waiting time, and appointment attendance.
- Analyse attendance patterns by appointment day and time.

### Stage 4: KPI Analysis
Calculate and compare the proposed KPIs across relevant patient and appointment segments.

### Stage 5: Insight Development
Identify meaningful patterns and relationships that may help HealthConnect understand the causes or characteristics associated with missed appointments.

### Stage 6: Recommendations
Translate the findings into practical, data-driven recommendations that HealthConnect can consider for improving appointment attendance and patient engagement.

## 12. Assumptions, Limitations, Risks and Dependencies

### Assumptions

- The dataset contains fictional and anonymized appointment records.
- `appointment_outcome` is assumed to represent the final outcome of each scheduled appointment.
- Each `appointment_id` represents a unique appointment.
- Multiple appointments may belong to the same patient because `patient_id` is not unique.

### Limitations

- The dataset may not contain all factors that influence whether a patient attends an appointment.
- Missing values are present in reminder channel, distance to clinic, and waiting time.
- The dataset is fictional and may not fully represent real-world patient behaviour.
- Observed relationships between variables should not automatically be interpreted as causal relationships.

### Risks

- Incorrect treatment of cancelled appointments could distort no-show analysis.
- Missing data could affect comparisons between patient groups.
- Small groups within certain categories may produce less reliable comparisons.
- Analysis based on historical patterns may not necessarily predict future patient behaviour.

### Dependencies

- The HealthConnect Data Dictionary is required to correctly interpret the variables.
- Further analysis depends on validating the meaning and consistency of the available data.
- Subsequent recommendations will depend on the patterns identified during detailed analysis.

## 13. Week 4 Project Summary

### Problem Addressed

The Data Analytics track will investigate appointment attendance and no-show patterns at HealthConnect Clinic. The aim is to identify factors and patterns that may help the clinic reduce missed appointments and improve the use of available appointment slots.

### Resources Used

The analysis is based on the HealthConnect Appointment Dataset and the HealthConnect Data Dictionary. The dataset contains 5,000 appointment records and 18 variables covering patient demographics, appointment details, booking information, previous appointment behaviour, reminders, distance to the clinic, waiting time, and appointment outcomes.

### Key Observations

The initial assessment identified three appointment outcomes: No-Show, Attended, and Cancelled. The dataset contains no completely duplicated records, and all appointment IDs are unique.

Missing values were identified in `reminder_channel`, `distance_to_clinic_km`, and `waiting_time_minutes`. The reminder channel requires particular attention because its missing values may be related to whether a reminder was sent.

The dataset also contains repeated patients, with 1,696 unique patient IDs across 5,000 appointments, making previous appointment and no-show history potentially relevant to the analysis.

### Proposed Approach

The next stage will involve preparing the data for analysis, investigating appointment attendance and no-show patterns, analysing the relationship between reminders and outcomes, and examining demographic, behavioural, scheduling, and logistical factors that may be associated with missed appointments.

The proposed KPIs include No-Show Rate, Attendance Rate, Reminder Effectiveness Rate, No-Show Rate by Appointment Type, and Repeat No-Show Rate.

### Key Considerations

The analysis will need to account for missing values and carefully distinguish cancelled appointments from no-shows. Because the dataset is fictional and observational, relationships identified during analysis should not automatically be interpreted as causal.

### Proposed Week 5 Focus

The proposed focus for Week 5 is to conduct deeper exploratory analysis of appointment attendance and no-show patterns, calculate and analyse the proposed KPIs, investigate relationships between relevant variables and appointment outcomes, and develop data-driven insights that can inform strategies for reducing missed appointments.

## 14. Conclusion

Week 4 established the foundation for the HealthConnect Data Analytics project by defining the business problem, reviewing the available appointment data, assessing initial data quality, identifying relevant variables, defining business questions, and proposing potential KPIs.

The initial assessment indicates that the dataset provides useful information for investigating appointment attendance and no-show patterns. The findings and proposed analytical approach will guide the detailed analysis and insight development in the next stage of the project.